Fix up udev rules to use uaccess and QHY-specific paths for fxload and
firmware
--- lib/udev/rules.d/85-qhyccd.rules	2019-02-19 00:17:08.954292732 +0000
+++ lib/udev/rules.d/85-qhyccd.rules	2019-02-19 00:23:17.066794879 +0000
@@ -18,137 +18,137 @@
 # 1. Pre-enumerated device IDs:
 #===============================================================================
 #Van Ooijen Technische Informatica
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081a", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="0818", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2972", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY8.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081a", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="0818", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2972", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY8.HEX -D $env{DEVNAME}"
 
 # Orion SSAG
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1856", ATTRS{idProduct}=="0011", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1856", ATTRS{idProduct}=="0011", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
 
 # CCD-Labs QGuider
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c7c", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c7c", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
 
 # Other
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="8613", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5LOADER.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="8613", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5LOADER.HEX -D $env{DEVNAME}"
 ACTION=="add", SUBSYSTEM=="usb", SYSFS{idVendor}=="0547", SYSFS{idProduct}=="1002", RUN+="/bin/sleep 5"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2970", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY2PRO.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2970", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY2PRO.HEX -D $env{DEVNAME}"
 
 # QHY
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="b618", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/IMG0H.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0412", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY2.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="296c", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="296c", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0901", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1002", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0259", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY6.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6000", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY8.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1000", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY10.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1200", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY12.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2850", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/IMG2P.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0920", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5II.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8300", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY9S.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6940", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY22.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6740", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY21.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8140", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY23.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1110", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY11.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1610", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY16000.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6002", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY8PRO.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0930", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/miniCam5.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4022", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY7.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1500", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY15.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1670", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY28.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2950", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY29.HEX -D $env{DEVNAME}"
-
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6004", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/QHY8L.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="b285", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IMG2S.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8310", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IC8300.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1650", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/QHY27.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1620", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IC16200A.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8320", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IC90A.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1630", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IC16803.HEX -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0940", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/POLEMASTER.HEX -D $env{DEVNAME}"
-
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0174", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III174.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0178", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III178.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0185", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III185.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0224", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III224.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0290", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III290.img -D $env{DEVNAME}"
-
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4040", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY4040.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2020", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY2020.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4200", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY42.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4202", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY42PRO.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="C550", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY550.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="C294", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY294.img -D $env{DEVNAME}"
-
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c128", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY128.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c163", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY163.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c165", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY165.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c167", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY168.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c174", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY174.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c178", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY178.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c183", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY183.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c224", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY224.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c247", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY247.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c290", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY290.img -D $env{DEVNAME}"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c367", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY367.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="b618", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/IMG0H.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0412", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY2.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="296c", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="296c", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0901", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1002", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0259", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY6.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6000", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY8.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1000", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY10.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1200", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY12.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2850", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/IMG2P.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0920", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5II.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8300", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY9S.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6940", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY22.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6740", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY21.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8140", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY23.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1110", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY11.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1610", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY16000.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6002", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY8PRO.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0930", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/miniCam5.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4022", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY7.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1500", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY15.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1670", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY28.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2950", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY29.HEX -D $env{DEVNAME}"
+
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6004", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/QHY8L.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="b285", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IMG2S.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8310", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IC8300.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1650", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/QHY27.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1620", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IC16200A.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8320", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IC90A.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1630", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IC16803.HEX -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0940", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/POLEMASTER.HEX -D $env{DEVNAME}"
+
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0174", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III174.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0178", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III178.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0185", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III185.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0224", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III224.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0290", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III290.img -D $env{DEVNAME}"
+
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4040", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY4040.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2020", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY2020.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4200", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY42.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4202", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY42PRO.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="C550", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY550.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="C294", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY294.img -D $env{DEVNAME}"
+
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c128", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY128.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c163", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY163.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c165", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY165.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c167", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY168.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c174", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY174.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c178", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY178.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c183", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY183.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c224", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY224.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c247", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY247.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c290", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY290.img -D $env{DEVNAME}"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c367", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY367.img -D $env{DEVNAME}"
 #===============================================================================
 # 2. Post-enumeration device IDs:
 #===============================================================================
 # Other Vendor
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="a618", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081d", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081e", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="296d", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2971", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2972", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1856", ATTRS{idProduct}=="0011", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c7c", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="8613", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", MODE="0666", GROUP="plugdev"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="a618", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081d", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081e", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="296d", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2971", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2972", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1856", ATTRS{idProduct}=="0011", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c7c", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="8613", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", TAG+="uaccess"
 
 # QHYCCD
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2972", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1001", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2851", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="a285", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0921", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8301", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8311", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8321", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6941", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6741", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8141", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1111", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1611", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6003", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0931", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4023", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1501", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1651", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1671", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2951", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1621", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8321", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0291", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0225", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0179", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0175", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0186", MODE="0666", GROUP="plugdev"
-
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c129", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c164", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c166", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c168", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c175", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c179", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c184", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c225", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c248", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c291", MODE="0666", GROUP="plugdev"
-ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c368", MODE="0666", GROUP="plugdev"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2972", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1001", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2851", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="a285", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0921", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8301", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8311", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8321", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6941", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6741", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8141", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1111", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1611", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6003", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0931", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4023", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1501", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1651", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1671", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2951", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1621", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8321", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0291", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0225", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0179", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0175", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0186", TAG+="uaccess"
+
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c129", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c164", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c166", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c168", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c175", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c179", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c184", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c225", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c248", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c291", TAG+="uaccess"
+ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c368", TAG+="uaccess"
 #
 #===============================================================================
 # Original stuff from QHYCCD company.
@@ -162,75 +162,75 @@
 #LABEL="qhy_vidpid_test"
 
 #Van Ooijen Technische Informatica
-#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081a", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
-#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="0818", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
-#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2972", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY8.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081a", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
+#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="0818", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
+#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2972", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY8.HEX -D $env{DEVNAME}"
 
 # Orion SSAG
-#ATTRS{idVendor}=="1856", ATTRS{idProduct}=="0011", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
+#ATTRS{idVendor}=="1856", ATTRS{idProduct}=="0011", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
 
 # CCD-Labs QGuider
-#ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c7c", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
+#ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c7c", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
 
 # Other
-#ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="8613", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME} -s /usr/local/lib/qhy/firmware/QHY5LOADER.HEX"
-#ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5LOADER.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="8613", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME} -s /lib/firmware/qhy/QHY5LOADER.HEX"
+#ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5LOADER.HEX -D $env{DEVNAME}"
 #SYSFS{idVendor}=="0547", SYSFS{idProduct}=="1002", RUN+="/bin/sleep 5"
-#ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2970", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY2PRO.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2970", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY2PRO.HEX -D $env{DEVNAME}"
 
 # Stock QHY
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="b618", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/IMG0H.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0412", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY2.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="296c", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="296c", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0901", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1002", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0259", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY6.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6000", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY8.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1000", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY10.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1200", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY12.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2850", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/IMG2P.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0920", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY5II.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8300", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY9S.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6940", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY22.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6740", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY21.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8140", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY23.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1110", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY11.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1610", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY16000.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6002", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY8PRO.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0930", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/miniCam5.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4022", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY7.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1500", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY15.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1670", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY28.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2950", RUN+="/sbin/fxload -t fx2 -I /usr/local/lib/qhy/firmware/QHY29.HEX -D $env{DEVNAME}"
-
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6004", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/QHY8L.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="b285", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IMG2S.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8310", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IC8300.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1650", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/QHY27.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1620", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IC16200A.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8320", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IC90A.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1630", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/IC16803.HEX -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0940", RUN+="/sbin/fxload -t fx2lp -I /usr/local/lib/qhy/firmware/POLEMASTER.HEX -D $env{DEVNAME}"
-
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0174", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III174.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0178", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III178.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0185", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III185.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0224", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III224.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0290", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY5III290.img -D $env{DEVNAME}"
-
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c128", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY128.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c163", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY163.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c165", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY165.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c167", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY168.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c174", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY174.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c178", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY178.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c183", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY183.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c224", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY224.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c247", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY247.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c290", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY290.img -D $env{DEVNAME}"
-#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c367", RUN+="/sbin/fxload -t fx3 -I /usr/local/lib/qhy/firmware/QHY367.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="b618", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/IMG0H.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0412", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY2.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="296c", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="296c", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0901", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1002", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0259", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY6.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6000", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY8.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1000", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY10.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1200", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY12.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2850", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/IMG2P.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0920", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY5II.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8300", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY9S.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6940", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY22.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6740", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY21.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8140", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY23.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1110", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY11.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1610", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY16000.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6002", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY8PRO.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0930", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/miniCam5.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4022", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY7.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1500", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY15.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1670", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY28.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2950", RUN+="/sbin/qhyfxload -t fx2 -I /lib/firmware/qhy/QHY29.HEX -D $env{DEVNAME}"
+
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6004", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/QHY8L.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="b285", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IMG2S.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8310", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IC8300.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1650", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/QHY27.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1620", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IC16200A.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8320", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IC90A.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1630", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/IC16803.HEX -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0940", RUN+="/sbin/qhyfxload -t fx2lp -I /lib/firmware/qhy/POLEMASTER.HEX -D $env{DEVNAME}"
+
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0174", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III174.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0178", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III178.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0185", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III185.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0224", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III224.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0290", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY5III290.img -D $env{DEVNAME}"
+
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c128", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY128.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c163", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY163.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c165", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY165.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c167", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY168.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c174", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY174.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c178", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY178.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c183", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY183.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c224", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY224.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c247", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY247.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c290", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY290.img -D $env{DEVNAME}"
+#ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c367", RUN+="/sbin/qhyfxload -t fx3 -I /lib/firmware/qhy/QHY367.img -D $env{DEVNAME}"
 
 #LABEL="qhy_vidpid_end"
 
@@ -244,58 +244,58 @@
 #LABEL="qhy-vipi"
 
 # Other Vendor
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="a618", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081d", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081e", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="296d", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2971", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2972", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1856", ATTRS{idProduct}=="0011", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c7c", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="8613", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", MODE="0666"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="a618", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081d", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="081e", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="296d", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2971", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="2972", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1856", ATTRS{idProduct}=="0011", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1781", ATTRS{idProduct}=="0c7c", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="04b4", ATTRS{idProduct}=="8613", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="0547", ATTRS{idProduct}=="1002", TAG+="uaccess"
 
 # Stock QHY
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2972", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1001", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2851", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="a285", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0921", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8301", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8311", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8321", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6941", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6741", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8141", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1111", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1611", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6003", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0931", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4023", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1501", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1651", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1671", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2951", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1621", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8321", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0291", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c291", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0225", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c225", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0179", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0175", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0186", MODE="0666"
-
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c129", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c164", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c166", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c168", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c175", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c179", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c184", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c225", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c248", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c291", MODE="0666"
-#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c368", MODE="0666"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2972", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1001", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2851", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="a285", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0921", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8301", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8311", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8321", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6941", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6741", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8141", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1111", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1611", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="6003", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0931", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="4023", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1501", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1651", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1671", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="2951", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="1621", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="8321", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0291", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c291", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0225", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c225", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0179", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0175", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="0186", TAG+="uaccess"
+
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c129", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c164", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c166", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c168", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c175", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c179", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c184", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c225", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c248", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c291", TAG+="uaccess"
+#SUBSYSTEMS=="usb", ATTRS{idVendor}=="1618", ATTRS{idProduct}=="c368", TAG+="uaccess"
 
 #LABEL="qhy-rules-end"
